adapters/rhythmone/rhythmone.go

```diff
@@ -0,0 +1,218 @@
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("bad server response: %d. ", err),
+		}}
+	}
+
+	var errs []error
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	return bidResponse, errs
+}
+
+func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			if imp.Banner != nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			}
+			return mediaType
+		}
+	}
+	return mediaType
+}
+
+func NewRhythmoneAdapter(config *adapters.HTTPAdapterConfig, endpoint string) *RhythmoneAdapter {
+	return NewRhythmoneBidder(adapters.NewHTTPAdapter(config).Client, endpoint)
+}
+
+func NewRhythmoneBidder(client *http.Client, endpoint string) *RhythmoneAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+	return &RhythmoneAdapter{
+		http:     a,
+		endPoint: endpoint,
+	}
+}
+
+func (a *RhythmoneAdapter) preProcess(req *openrtb.BidRequest, errors []error) (*openrtb.BidRequest, string, []error) {
+	numRequests := len(req.Imp)
+
+	requestImpCopy := req.Imp
+	var uri string = ""
+	for i := 0; i < numRequests; i++ {
+		imp := requestImpCopy[i]
+
+		if imp.Audio != nil {
+			continue
+		}
+		if imp.Native != nil {
+			continue
+		}
+		if imp.Banner != nil || imp.Video != nil {
+
+			var bidderExt adapters.ExtImpBidder
+			err := json.Unmarshal(imp.Ext, &bidderExt)
+
+			if err != nil {
+				err = &errortypes.BadInput{
+					Message: fmt.Sprintf("ext data not provided in imp id=%s. Abort all Request", imp.ID),
+				}
+				errors = append(errors, err)
+				return nil, "", errors
+			}
+			var rhythmoneExt openrtb_ext.ExtImpRhythmone
+			err = json.Unmarshal(bidderExt.Bidder, &rhythmoneExt)
+			if err != nil {
+				err = &errortypes.BadInput{
+					Message: fmt.Sprintf("placementId | zone | path not provided in imp id=%s. Abort all Request", imp.ID),
+				}
+				errors = append(errors, err)
+				return nil, "", errors
+			}
+
+			if rhythmoneExt.PlacementId == "" || rhythmoneExt.Zone == "" || rhythmoneExt.Path == "" {
```

Comment:
This constraint should go in your `static/bidder-params/rhythmone.json` schema as a `minlength`: https://json-schema.org/understanding-json-schema/reference/string.html#length

Then you can remove this logic here. If the params don't validate against that schema, your adapter won't be called.